nav2_safety_monitor/README.md

```diff
@@ -0,0 +1,68 @@
+# Nav2 Safety Monitor
+
+`nav2_safety_monitor` is a package containing an implementation of a collision monitor for use of a 2D laser scanner, sonars, and contact/bump sensors, the main 3 "safety" sensors on a robot. More can be easily added by creating an instance of the SafetySensor abstract class. 
+
+It looks at the range values of obstacles in a configurable box. If the laser has returns in this box, the robot will be commanded to stop until the obstacle is gone or a timeout is achieved. For contact sensors, it looks for the boolean if in contact to stop. For sonars/range sensors, it is dependent on the non-zero ranges whether it falls into the safety (slow down) region or the collision region. Parameters for all can be seen in the example below.
+
+This may be an issue if you plan to get very close to obstacles and have a large collision box configuration. For applications where docking/interacting with items is required, a service is exposed so that the monitor can be toggled on or off as needed for specific maneuvers and get that state from an application level. 
+
+```
+Fully described parameter for the process this belongs to
+
+process:
+  process:
+    ros__parameters:
+      sensors: [laser, sonar1, sonar2, bump]
+      laser:
+        type: laser
+        topic: scan
+      sonar1:
+        type: sonar
+        topic: sonar_1
+      sonar2:
+        type: sonar
+        topic: sonar_2
+      bump:
+        type: contact
+        topic: bump
+      collision_zone: [0.1, 0.1, 0.0, 0.0]
+      safety_zone: [0.3, 0.3, 0.0, 0.0]
+      safety_timeout: 120.0
+      scan_timeout: 1.0
+      sonar_timeout: 1.0
+      global_frame: base_link
+      sonar_safety_range: 0.3
+      sonar_collision_range: 0.1
+```
+
+Example useage:
+
+```
+#include <nav2_safety_monitor/nav2_safety_monitor.hpp>
+
+auto node = ...
+
+nav2_safety_monitor::SafetyMonitor monitor(node);
+action_in_progress = true;
+
+while (rclcpp.ok() && action_in_progress) {
+
+  // do some stuff with a planner, controller, actuators, ... 
+
+  if (!monitor.isActive()) {
+    monitor.activate();
+  }
+
+  monitor.process();
```

Comment:
Most of the nav stack takes the approach of doing lots of work in the message callbacks and almost none in the main loop. You've taken the opposite approach; the callbacks do no work except record the message, and all the work happens on the call to `process`.

I'm curious to hear why you've taken this approach, and if you feel it should be more generally applied to the stack.

Reply:
Message callbacks for lasers can come in at 40hz+, anything laser related shouldn't be happening in the callback in my opinion unless it can keep up. Even then if you're only processing the state of that information at 10hz in the case of a planner or 20hz in the case of a controller, those additional 50% of computations are wasted. 

This is something I always do with high rate sensors like lasers, sonars, and depth cameras.